src/utility/copy.cpp

```diff
@@ -208,32 +213,103 @@ CheckRewritten(List *rewritten) {
 	}
 }
 
-bool
+static bool
 CheckPrefix(const char *str, const char *prefix) {
-	while (*prefix) {
-		if (*prefix++ != *str++) {
+	return strncmp(str, prefix, strlen(prefix)) == 0;
+}
+
+static bool
+IsAllowedStatement(CopyStmt *stmt, bool throw_error = false) {
+	int elevel = throw_error ? ERROR : DEBUG4;
+
+	if (stmt->relation) {
+		Relation rel = table_openrv(stmt->relation, AccessShareLock);
+		bool is_duckdb_table = pgduckdb::IsDuckdbTable(rel);
+		bool is_catalog_table = pgduckdb::IsCatalogTable(rel);
+		table_close(rel, NoLock);
+		if (is_catalog_table) {
+			elog(elevel, "DuckDB does not support querying PG catalog tables");
 			return false;
 		}
+
+		if (stmt->is_from && !is_duckdb_table) {
+			elog(elevel, "pg_duckdb does not support COPY ... FROM ... yet for Postgres tables");
+			return false;
+		}
+	}
+
+	if (stmt->filename == NULL) {
+		elog(elevel, "COPY ... TO STDOUT/FROM STDIN is not supported by DuckDB");
+		return false;
 	}
+
 	return true;
 }
-
-const char *
-MakeDuckdbCopyQuery(PlannedStmt *pstmt, const char *query_string, struct QueryEnvironment *query_env) {
-	CopyStmt *copy_stmt = (CopyStmt *)pstmt->utilityStmt;
-	if (!copy_stmt->filename) {
-		return nullptr;
+static bool
+ContainsDuckdbCopyOption(CopyStmt *stmt) {
+	foreach_node(DefElem, defel, stmt->options) {
+		if (strcmp(defel->defname, "format") == 0) {
```

Comment:
Our code is littered with `strcmp(str1, str2) == 0`, should we add an `IsEq` helper? (not in this PR ofc)

Reply:
done